qa/tasks/cephfs/filesystem.py

```diff
@@ -561,7 +562,34 @@ def __init__(self, ctx, fs_config={}, fscid=None, name=None, create=False, clust
         self.data_pool_name = None
         self.data_pools = None
         self.fs_config = fs_config
-        self.ec_profile = fs_config.get('ec_profile')
+        if self.fs_config:
+            log.debug(f"Filesystem init - using passed config={self.fs_config} for fsname={self.name}")
+        elif self.name:
+            # Construct fs_config from config stored in context.
+            log.debug(f"Filesystem init - using context cephfs_config={ctx.cephfs_config} for fsname={self.name}")
+            log.debug(f"Filesystem init - using context subvols={ctx.subvols} for fsname={self.name}")
+            # There are generic fs options which should apply to all the filesystems and each
+            # filesystem can have extra options or can override the generic ones. The
+            # temp_cephfs_config has generic fs options and the fs_configs has fs specific options
+            # after it's popped from temp_cephfs_config. The for loop iterates through fs specific
+            # option and merges them with generic fs option and construct the final list for the
+            # particular fs.
+            temp_cephfs_config = deepcopy(ctx.cephfs_config)
+            subvols = deepcopy(ctx.subvols)
+            fs_configs =  temp_cephfs_config.pop('fs', [{'name': 'cephfs'}])
+            for fs_conf in fs_configs:
+                assert isinstance(fs_conf, dict)
+                fs_name = fs_conf.pop('name')
+                if fs_name == self.name:
+                    misc.deep_merge(temp_cephfs_config, fs_conf)
+                    if subvols:
+                        misc.deep_merge(temp_cephfs_config, {'subvols': subvols})
+                    self.fs_config = temp_cephfs_config
+            log.debug(f"Filesystem init - constructed from context - cephfs_config {self.fs_config} for {self.name}")
```

Comment:
If we are deep merging only if `fs_name == self.name`, do we really need to run the `for` loop over `fs_configs` ?

Can't we just test if `self.name` is in the set of keys under `fs_configs` and then deep merge ?

Reply:
I didn't get it. The `fs` would hold the list of fs_config dictionaries with fs name as one of they key as below. You would anyway need to traverse the list to identify if the `fs_name` is present.

```
ceph:
  cephfs:
    allow_referent_inodes: true
    max_mds: 2
    fs:
      - name: a
        standby_replay: true
        max_mds: 3
      - name: b
        standby_replay: false
        max_mds:1
```

The above would translate to:

```
fs_configs = [{'name':'a', 'standby_replay':'true', 'max_mds':'3'},{'name':'b', 'standby_replay':'false', 'max_mds':'1'}]
```